File: python/reorder_routes_to_make_all_paths_lead_to_the_city_zero.py
class Solution:
    def minReorder(self, n: int, connections: list[list[int]]) -> int:
        """time complexity: O(n), space complexity: O(n)"""
        edges: set[tuple[int, int]] = {(fr, to) for fr, to in connections}
        graph: dict[int, list[int]] = {}
        visited: set[int] = set()
        changes: list[int] = [0]
        for city in range(n):
            if city not in graph:
                graph[city] = []
        for fr, to in connections:
            if fr not in graph:
                graph[fr] = [to]
            else:
                graph[fr].append(to)
            if to not in graph:
                graph[to] = [fr]
            else:
                graph[to].append(fr)
        visited.add(0)
        self.dfs(0, edges, graph, visited, changes)
        return changes[0]

    def dfs(
        self,
        city: int,
        edges: dict[int, int],
        graph: dict[int, list[int]],
        visited: set[int],
        changes: list[int],
    ) -> None:
        for neighbor in graph[city]:
            if neighbor in visited:
                continue
            if (neighbor, city) not in edges:
                changes[0] += 1
            visited.add(neighbor)
            self.dfs(neighbor, edges, graph, visited, changes)


def test_minReorder_case_1():
    # arrange
    n: int = 6
    connections: list[list[int]] = [[0, 1], [1, 3], [2, 3], [4, 0], [4, 5]]
    expected: int = 3

    # act
    solution = Solution()
    actual = solution.minReorder(n, connections)

    # assert
    assert expected == actual


def test_minReorder_case_2():
    # arrange
    n: int = 5
    connections: list[list[int]] = [[1, 0], [1, 2], [3, 2], [3, 4]]
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.minReorder(n, connections)

    # assert
    assert expected == actual

File: rust/src/code/reorder_routes_to_make_all_paths_lead_to_the_city_zero.rs
struct Solution {}

use std::collections::{HashMap, HashSet};

impl Solution {
    // NOTE: time complexity O(n)
    pub fn min_reorder(n: i32, connections: Vec<Vec<i32>>) -> i32 {
        let mut changes: i32 = 0;
        let mut edges: HashSet<(i32, i32)> = HashSet::new();
        let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
        let mut visited: HashSet<i32> = HashSet::new();
        for city in 0..n {
            graph.insert(city, vec![]);
        }
        for connection in connections {
            edges.insert((connection[0], connection[1]));
            graph
                .entry(connection[0])
                .and_modify(|x| x.push(connection[1]))
                .or_insert(vec![connection[1]]);
            graph
                .entry(connection[1])
                .and_modify(|x| x.push(connection[0]))
                .or_insert(vec![connection[0]]);
        }

        visited.insert(0);
        Self::dfs(0, &edges, &graph, &mut visited, &mut changes);
        changes
    }

    fn dfs(
        city: i32,
        edges: &HashSet<(i32, i32)>,
        graph: &HashMap<i32, Vec<i32>>,
        visited: &mut HashSet<i32>,
        changes: &mut i32,
    ) {
        for neighbor in graph[&city].iter() {
            if visited.contains(neighbor) {
                continue;
            }
            if !edges.contains(&(*neighbor, city)) {
                *changes += 1;
            }
            visited.insert(*neighbor);
            Self::dfs(*neighbor, edges, graph, visited, changes);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_min_reorder_case_1() {
        // arrange
        let n: i32 = 6;
        let connections: Vec<Vec<i32>> =
            vec![vec![0, 1], vec![1, 3], vec![2, 3], vec![4, 0], vec![4, 5]];
        let expected: i32 = 3;

        // act
        let actual = Solution::min_reorder(n, connections);

        // assert
        assert_eq!(expected, actual);
    }
    #[test]
    fn test_min_reorder_case_2() {
        // arrange
        let n: i32 = 5;
        let connections: Vec<Vec<i32>> = vec![vec![1, 0], vec![1, 2], vec![3, 2], vec![3, 4]];
        let expected: i32 = 2;

        // act
        let actual = Solution::min_reorder(n, connections);

        // assert
        assert_eq!(expected, actual);
    }
}
